perf/rdm_bw.c

```diff
@@ -0,0 +1,471 @@
+	int size, align_size;
+	char *s_buf, *r_buf;
+	double t_start = 0.0, t_end = 0.0, t = 0.0;
+	int op, ret;
+	buf_desc_t lbuf_desc;
+	ssize_t fi_rc;
+
+	FT_Init(&argc, &argv);
+	FT_Rank(&myid);
+	FT_Job_size(&numprocs);
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return -1;
+
+	while ((op = getopt(argc, argv, "f:h")) != -1) {
+		switch (op) {			
+		case 'f':
+			hints->fabric_attr->prov_name = strdup(optarg);
+			break;
+		case '?':
+		case 'h':
+			print_usage();
+			return -1;
+		}
+	}
+
+	hints->ep_attr->type	= FI_EP_RDM;
+	hints->caps		= FI_MSG | FI_RMA;
+	hints->mode		= ~0;
+	hints->domain_attr->mr_mode = FI_MR_BASIC;
+
+	if(numprocs != 2) {
+		if(myid == 0) {
+			fprintf(stderr, "This test requires exactly two processes\n");
+		}
+		FT_Finalize();
+		return -1;
+	}
+
+	/* Fabric initialization */
+	ret = init_fabric();
+	if(ret) {
+		fprintf(stderr, "Problem in fabric initialization\n");
+		return ret;
+	}
+
+	ret = init_av();
+	if(ret) {
+		fprintf(stderr, "Problem in AV initialization\n");
+		return ret;
+	}
+
+	/* Data initialization */
+	align_size = getpagesize();
+	assert(align_size <= MAX_ALIGNMENT);
+
+	s_buf = (char *) (((unsigned long) s_buf_original + (align_size - 1)) /
+				align_size * align_size);
+	r_buf = (char *) (((unsigned long) r_buf_original + (align_size - 1)) /
+				align_size * align_size);
+
+	ret = fi_mr_reg(dom, r_buf, MYBUFSIZE, FI_REMOTE_WRITE, 0, 0, 0, &r_mr, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_mr_reg", ret);
+		return -1;
+	}
+
+	lbuf_desc.addr = (uint64_t)r_buf;
+	lbuf_desc.key = fi_mr_key(r_mr);
+
+	rbuf_descs = (buf_desc_t *)malloc(numprocs * sizeof(buf_desc_t));
+
+	/* Distribute memory keys */
+	FT_Allgather(&lbuf_desc, sizeof(lbuf_desc), rbuf_descs);
+
+	ret = fi_mr_reg(dom, s_buf, MYBUFSIZE, FI_WRITE, 0, 0, 0, &l_mr, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_mr_reg", ret);
+		return -1;
+	}
+
+	if (myid == 0) {
+		fprintf(stdout, HEADER);
+		fprintf(stdout, "%-*s%*s\n", 10, "# Size", FIELD_WIDTH,
+				"Bandwidth (MB/s)");
+		fflush(stdout);
+	}
+
+	/* Bandwidth test */
+	for (size = 1; size <= MAX_MSG_SIZE; size *= 2) {
+		/* touch the data */
+		for (i = 0; i < size; i++) {
+			s_buf[i] = 'a';
+			r_buf[i] = 'b';
+		}
+
+		if (size > large_message_size) {
+			loop = loop_large;
+			skip = skip_large;
+			window_size = window_size_large;
+		}
+
+		if (myid == 0) {
+			peer = 1;
+
+			for (i = 0; i < loop + skip; i++) {
+				if (i == skip) {
+					t_start = get_time();
+				}
+
+				for (j = 0; j < window_size; j++) {
+					fi_rc = fi_write(ep, s_buf, size, l_mr,
+							fi_addrs[peer],
+							rbuf_descs[peer].addr,
+							rbuf_descs[peer].key,
+							(void *)(intptr_t)j);
+					if (fi_rc) {
+						FT_PRINTERR("fi_write", fi_rc);
+						return fi_rc;
+					}
+				}
+
+				wait_for_data_completion(scq, window_size);
+			}
+
+			fi_rc = fi_send(ep, s_buf, 4, NULL,
+					fi_addrs[peer],
+					NULL);
+			assert(!fi_rc);
+			fi_rc = fi_recv(ep, s_buf, 4, NULL,
+					fi_addrs[peer],
+					NULL);
```

Comment:
fi_recv() replaces a call to MPI_Recv() here.  MPI_Recv() takes the rank 
of the peer to receive from as an argument.  I think this maps well to a 
directed receive.

-Zach

On 06/19/2015 11:51 AM, Sean Hefty wrote:
>
> In perf/rdm_bw.c 
> <https://github.com/ofiwg/fabtests/pull/279#discussion_r32847153>:
>
> > +					if (fi_rc) {
> > +						FT_PRINTERR("fi_write", fi_rc);
> > +						return fi_rc;
> > +					}
> > +				}
> > +
> > +				wait_for_data_completion(scq, window_size);
> > +			}
> > +
> > +			fi_rc = fi_send(ep, s_buf, 4, NULL,
> > +					fi_addrs[peer],
> > +					NULL);
> > +			assert(!fi_rc);
> > +			fi_rc = fi_recv(ep, s_buf, 4, NULL,
> > +					fi_addrs[peer],
> > +					NULL);
> Does FI_DIRECTED_RECV imply FI_RECV? I'm looking at using: (FI_MSG | 
> FI_DIRECTED_RECV), or (FI_SEND | FI_DIRECTED_RECV)
> FI_MSG by itself implies both send and receive. So, FI_MSG | 
> FI_DIRECTED_RECV should be fine. Note that the code doesn't actually 
> need directed receive, since it only supports one peer. Adding 
> directed receive is fine, but may cause it not to work with some 
> providers. I'm not as concerned with this test, versus whether this is 
> feature will be needed for future tests, and we want some level of 
> consistency here. If these tests are ports of the OSU benchmarks, I 
> would match the intent of those tests.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ofiwg/fabtests/pull/279/files#r32847153>.
>